wasp/benchspy/resources.go

```diff
@@ -0,0 +1,235 @@
+package benchspy
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/pkg/errors"
+	tc "github.com/testcontainers/testcontainers-go"
+	"golang.org/x/sync/errgroup"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type ExecutionEnvironment string
+
+const (
+	ExecutionEnvironment_Docker                  ExecutionEnvironment = "docker"
+	ExecutionEnvironment_k8sExecutionEnvironment ExecutionEnvironment = "k8s"
+)
+
+type ResourceReporter struct {
+	// either k8s or docker
+	ExecutionEnvironment ExecutionEnvironment `json:"execution_environment"`
+
+	// AUT metrics
+	PodsResources      map[string]*PodResources    `json:"pods_resources"`
+	ContainerResources map[string]*DockerResources `json:"container_resources"`
+	// regex pattern to select the resources we want to fetch
+	ResourceSelectionPattern string `json:"resource_selection_pattern"`
+}
+
+type DockerResources struct {
+	NanoCPUs   int64
+	CpuShares  int64
+	Memory     int64
+	MemorySwap int64
+}
+
+type PodResources struct {
+	RequestsCPU    int64
+	RequestsMemory int64
+	LimitsCPU      int64
+	LimitsMemory   int64
+}
+
+func (r *ResourceReporter) FetchResources(ctx context.Context) error {
+	if r.ExecutionEnvironment == ExecutionEnvironment_Docker {
+		err := r.fetchDockerResources(ctx)
+		if err != nil {
+			return err
+		}
+	} else {
+		err := r.fetchK8sResources(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ResourceReporter) fetchK8sResources(ctx context.Context) error {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return errors.Wrapf(err, "failed to get in-cluster config, are you sure this is running in a k8s cluster?")
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create k8s clientset")
+	}
+
+	namespaceFile := "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	namespace, err := os.ReadFile(namespaceFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read namespace file %s", namespaceFile)
+	}
+
+	pods, err := clientset.CoreV1().Pods(string(namespace)).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to list pods in namespace %s", namespace)
+	}
+
+	r.PodsResources = make(map[string]*PodResources)
```

Comment:
K8s dependencies are a heavy burden, if all we need is just to execute one query can we do it without pulling `k8s` deps into WASP? If not possible then okay

Reply:
Oh, it's a separate module, but I believe we can still think about this later, transient deps

Reply:
Let's also think if it's that crucial, do users need to know if limits/requests have changed or only consumption of such resources?

Reply:
@skudasov was about to ask... what do we really care about? limit/resources or usage stats?

Reply:
also, I haven't added these k8s deps... I guess they were sucked in, when I added dependency on the CTF?